OneCommunityGlobal/HGNRest
# Shiwani Create a Leaderboard way to see when people are off for the week

# Description
<img width="910" alt="Screenshot 2023-12-18 at 6 10 55 PM" src="https://github.com/OneCommunityGlobal/HGNRest/assets/78906820/6531e12b-396c-4173-8447-ff4c274b98f9">


## Related PRS (if any):
Related PRs [PR1542](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1542)+614, [PR1081](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1081)

To test this backend PR you need to checkout the [#PR1722](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1722) frontend PR.

## Main changes explained:
- Added timeOffFrom and timeOffTill fields to userProfile.js under models
- Added conditions in reasonSchedulingController.js to check if the fields timeOffFrom and timeOffTill existed, if they didn't then the fields would be added with their dates. If the fields were present, then another condition would be checked if the currentDate is between the last timeOffFrom and timeOffTill dates and if it is true then only timeOffTill date will be changed else both the fields will be updated. This is because if the user is already in a time off and is preponing or postponing their time off then it would update only timeOffTill in this case.
- Modified aggregation pipeline of dashboardhelper.js, taskHelper.js, reportsHelper.js to include both the fields so that the date can be sent to the frontend. Added null conditions in the pipeline to check if the fields existed and if they didn't the date values would be null or else the respective date would be added to these two fields.
- timeOffFrom and timeOffTill fields are basically the last timeOff the user has requested since the current date is the date from when the user requests a time off till the sunday when the user would return which is why I am updating these two fields. All previously requested time off's are present in reason collection.

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log in as any user
5. follow instructions in frontend PR

## Screenshots or videos of changes:

https://github.com/OneCommunityGlobal/HGNRest/assets/78906820/3987db44-0827-46d7-9630-40cc9d1070ae



## Note:
- Please do also test on how the frontend looks when the current date does not fall between the last time off dates requested and also when the time off dates are preponed or postponed.
- You can also modify dates in mongoDB to test how the frontend looks for different cases.
